File: tests/java/tests-java-gradle-junit5-allure3-playwright/src/test/java/pages/HomePage.java
package pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import io.qameta.allure.Step;

import static com.microsoft.playwright.assertions.PlaywrightAssertions.assertThat;
import static api.AuthApiClient.login;

public class HomePage {

    private final Page page;
    public final Locator layout;
    public final Locator healthStatus;
    public final Locator itemsList;
    public final Locator welcomeMessage;
    public final Locator welcomePanel;
    public final Locator logoutButton;
    public final Locator deleteAccountButton;
    public final Locator header;

    public HomePage(Page page) {
        this.page = page;
        this.layout = page.getByTestId("multistack-layout");
        this.healthStatus = page.getByTestId("health-status");
        this.itemsList = page.getByTestId("items-list");
        this.welcomeMessage = page.getByTestId("welcome-message");
        this.welcomePanel = page.getByTestId("welcome-panel");
        this.logoutButton = page.getByTestId("logout-button");
        this.deleteAccountButton = page.getByTestId("delete-account-button");
        this.header = page.getByTestId("header");
    }

    @Step("Open home page")
    public HomePage open() {
        // Relative to context baseURL (path prefix on prod/stage). navigate("/") is
        // origin-absolute and drops /stack/.../frontend-.../.
        page.navigate("./");
        return shouldBeOpen();
    }

    @Step("Verify home layout is open")
    public HomePage shouldBeOpen() {
        layout.waitFor();
        return this;
    }

    @Step("Verify reference layout is mounted")
    public HomePage shouldShowLayout() {
        layout.waitFor();
        itemsList.waitFor();
        return this;
    }

    @Step("Verify home layout and health are mounted")
    public HomePage shouldShowLayoutAndHealth() {
        layout.waitFor();
        healthStatus.waitFor();
        return this;
    }

    @Step("Logout")
    public HomePage logout() {
        logoutButton.click();
        return this;
    }

    @Step("Reload home")
    public HomePage reload() {
        page.reload();
        return shouldBeOpen();
    }

    @Step("Accept delete-account confirm")
    public HomePage clickDeleteAccountAndConfirm() {
        page.onceDialog(dialog -> dialog.accept());
        deleteAccountButton.click();
        return this;
    }

    @Step("Cancel delete-account confirm")
    public HomePage clickDeleteAccountAndCancel() {
        page.onceDialog(dialog -> dialog.dismiss());
        deleteAccountButton.click();
        return this;
    }

    @Step("Open home page with local storage authentication")
    public HomePage openWithLocalStorageAuthentication(String username, String password) {
        return openWithLocalStorageAuth(login(username, password));
    }

    @Step("Seed localStorage auth token")
    public HomePage openWithLocalStorageAuth(String token) {
        page.navigate("login");
        var key = authTokenKey();
        page.evaluate(
                "arg => localStorage.setItem(arg.key, arg.token)",
                java.util.Map.of("key", key, "token", token));
        return open();
    }

    @Step("Verify session panel offers logout and delete account")
    public HomePage shouldShowSessionActions() {
        assertThat(logoutButton).isVisible();
        assertThat(logoutButton).containsText("Logout");
        assertThat(deleteAccountButton).isVisible();
        assertThat(deleteAccountButton).containsText("Delete account");
        return this;
    }

    @Step("Open home with a garbage auth token")
    public HomePage openWithInvalidToken() {
        return openWithLocalStorageAuth("invalid-token");
    }

    public String authTokenKey() {
        var key = (String) page.evaluate("""
                () => {
                  const m = location.pathname.match(/\\/(backend-[^/]+)\\//);
                  return m ? `authToken:${m[1]}` : 'authToken';
                }
                """);
        return key;
    }

    public String authToken() {
        return (String) page.evaluate("k => localStorage.getItem(k)", authTokenKey());
    }
}

File: tests/csharp/tests-csharp-xunit-api_request-playwright/Pages/HomePage.cs
using Allure.Net.Commons.Attributes;
using Api;
using Helpers;
using Microsoft.Playwright;

namespace Pages;

public sealed class HomePage
{
    private const string DeleteAccountConfirm = "Delete this account? This cannot be undone.";

    private readonly IPage _page;
    public readonly ILocator Layout;
    public readonly ILocator HealthStatus;
    public readonly ILocator ItemsList;
    public readonly ILocator WelcomeMessage;
    public readonly ILocator WelcomePanel;
    public readonly ILocator LogoutButton;
    public readonly ILocator DeleteAccountButton;
    public HeaderPage Header { get; set; } = null!;

    public HomePage(IPage page)
    {
        _page = page;
        Layout = page.GetByTestId("multistack-layout");
        HealthStatus = page.GetByTestId("health-status");
        ItemsList = page.GetByTestId("items-list");
        WelcomeMessage = page.GetByTestId("welcome-message");
        WelcomePanel = page.GetByTestId("welcome-panel");
        LogoutButton = page.GetByTestId("logout-button");
        DeleteAccountButton = page.GetByTestId("delete-account-button");
    }

    [AllureStep("Open home page")]
    public HomePage OpenPage()
    {
        Pw.Run(_page.GotoAsync("./"));
        return ShouldBeOpen();
    }

    [AllureStep("Verify home layout is open")]
    public HomePage ShouldBeOpen()
    {
        Pw.Run(Layout.WaitForAsync());
        return this;
    }

    [AllureStep("Verify home layout is mounted")]
    public HomePage ShouldShowLayout()
    {
        Pw.Run(Layout.WaitForAsync());
        Pw.Run(ItemsList.WaitForAsync());
        return this;
    }

    [AllureStep("Verify home layout and health are mounted")]
    public HomePage ShouldShowLayoutAndHealth()
    {
        Pw.Run(Layout.WaitForAsync());
        Pw.Run(HealthStatus.WaitForAsync());
        return this;
    }

    [AllureStep("Verify health and items finished loading")]
    public HomePage ShouldShowSettledHealthAndItems()
    {
        ShouldShowLayoutAndHealth();
        Pw.Run(ItemsList.WaitForAsync());
        Pw.Run(Assertions.Expect(HealthStatus).Not.ToContainTextAsync("Checking health"));
        Pw.Run(Assertions.Expect(ItemsList).Not.ToContainTextAsync("Loading items"));
        return this;
    }

    [AllureStep("Home layout panel is visible")]
    public ILocator LayoutPanel() => Layout;

    [AllureStep("Welcome panel is visible")]
    public ILocator WelcomePanelElement() => WelcomePanel;

    [AllureStep("Verify welcome panel stays hidden")]
    public HomePage ShouldHideWelcomePanel()
    {
        Pw.Run(Assertions.Expect(WelcomePanel).ToHaveAttributeAsync("hidden", ""));
        return this;
    }

    [AllureStep("Verify auth token was cleared from localStorage")]
    public HomePage ShouldClearAuthToken()
    {
        Pw.Run(_page.WaitForFunctionAsync(
            """
            () => {
              const m = location.pathname.match(/\/(backend-[^/]+)\//);
              const key = m ? `authToken:${m[1]}` : 'authToken';
              return localStorage.getItem(key) === null;
            }
            """));
        return this;
    }

    [AllureStep("Verify health status contains: {textFragment}")]
    public HomePage ShouldShowHealthText(string textFragment)
    {
        Pw.Run(Assertions.Expect(HealthStatus).ToContainTextAsync(textFragment));
        return this;
    }

    [AllureStep("Verify items list contains: {textFragment}")]
    public HomePage ShouldShowItemText(string textFragment)
    {
        Pw.Run(Assertions.Expect(ItemsList).ToContainTextAsync(textFragment));
        return this;
    }

    [AllureStep("Verify items panel shows a readable error: {textFragment}")]
    public HomePage ShouldShowItemsError(string textFragment)
    {
        Pw.Run(Assertions.Expect(ItemsList).ToContainTextAsync(textFragment));
        return this;
    }

    [AllureStep("Verify health panel shows a readable error: {textFragment}")]
    public HomePage ShouldShowHealthError(string textFragment)
    {
        Pw.Run(Assertions.Expect(HealthStatus).ToContainTextAsync(textFragment));
        return this;
    }

    [AllureStep("Verify welcome message: {message}")]
    public HomePage ShouldHaveWelcomeMessage(string message)
    {
        Pw.Run(WelcomePanel.WaitForAsync());
        Pw.Run(Assertions.Expect(WelcomeMessage).ToContainTextAsync(message));
        return this;
    }

    [AllureStep("Verify session panel offers logout and delete account")]
    public HomePage ShouldShowSessionActions()
    {
        Pw.Run(Assertions.Expect(LogoutButton).ToBeVisibleAsync());
        Pw.Run(Assertions.Expect(LogoutButton).ToContainTextAsync("Logout"));
        Pw.Run(Assertions.Expect(DeleteAccountButton).ToBeVisibleAsync());
        Pw.Run(Assertions.Expect(DeleteAccountButton).ToContainTextAsync("Delete account"));
        return this;
    }

    [AllureStep("Logout")]
    public LoginPage ClickLogoutButton()
    {
        Pw.Run(LogoutButton.ClickAsync());
        return new LoginPage(_page) { Header = Header };
    }

    [AllureStep("Accept delete-account confirm")]
    public LoginPage ClickDeleteAccountAndConfirm()
    {
        HandleDialog(accept: true);
        Pw.Run(DeleteAccountButton.ClickAsync());
        return new LoginPage(_page) { Header = Header };
    }

    [AllureStep("Cancel delete-account confirm")]
    public HomePage ClickDeleteAccountAndCancel()
    {
        HandleDialog(accept: false);
        Pw.Run(DeleteAccountButton.ClickAsync());
        return this;
    }

    [AllureStep("Verify auth token remains in localStorage")]
    public HomePage ShouldKeepAuthToken()
    {
        Assert.NotNull(AuthToken());
        return this;
    }

    [AllureStep("Open home page with local storage authentication")]
    public HomePage OpenPageWithLocalStorageAuthentication(string username, string password) =>
        OpenWithLocalStorageAuth(AuthApiClient.Login(username, password));

    [AllureStep("Seed localStorage auth token")]
    public HomePage OpenWithLocalStorageAuth(string token)
    {
        Pw.Run(_page.GotoAsync("login"));
        Pw.Run(_page.GetByTestId("login-form").WaitForAsync());
        var key = AuthTokenKey();
        Pw.Run(_page.EvaluateAsync(
            "arg => localStorage.setItem(arg.key, arg.token)",
            new { key, token }));
        return OpenPage();
    }

    [AllureStep("Open home with a garbage auth token")]
    public HomePage OpenPageWithInvalidToken() => OpenWithLocalStorageAuth("invalid-token");

    [AllureStep("Reload home")]
    public HomePage ReloadPage()
    {
        Pw.Run(_page.ReloadAsync());
        return ShouldBeOpen();
    }

    public string AuthTokenKey() =>
        Pw.Run(_page.EvaluateAsync<string>(
            """
            () => {
              const m = location.pathname.match(/\/(backend-[^/]+)\//);
              return m ? `authToken:${m[1]}` : 'authToken';
            }
            """)) ?? "authToken";

    public string? AuthToken() =>
        Pw.Run(_page.EvaluateAsync<string?>("k => localStorage.getItem(k)", AuthTokenKey()));

    private void HandleDialog(bool accept)
    {
        EventHandler<IDialog>? handler = null;
        handler = (_, dialog) =>
        {
            _page.Dialog -= handler!;
            if (dialog.Message != DeleteAccountConfirm)
            {
                throw new InvalidOperationException(
                    $"Confirm text: expected <{DeleteAccountConfirm}> but was <{dialog.Message}>");
            }

            Pw.Run(accept ? dialog.AcceptAsync() : dialog.DismissAsync());
        };
        _page.Dialog += handler;
    }
}
